### ISSUE TITLE
`awesome.kill` doesn't work on kill shell processes' subprocesses

### ISSUE THREAD

**Issue by doronbehar** [awesomeWM/awesome#2378]
Output of `awesome --version`:
```
awesome v4.2 (Human after all)
 • Compiled against Lua 5.3.4 (running with Lua 5.3)
 • D-Bus support: ✔
 • execinfo support: ✔
 • xcb-randr version: 1.5
 • LGI version: 0.9.2
```

**How to reproduce the issue:**

1. Add somewhere in your `rc.lua` this:

```lua
local pid = awful.spawn('/tmp/bin/awesome-test.sh')
awesome.connect_signal("exit", function(reason_restart)
	awesome.kill(pid, awesome.unix_signal['SIGTERM'])
end)
```

2. Put in `/tmp/bin/awesome-test.sh`:

```sh
#!/bin/sh

while inotifywait /tmp/awesome-test; do
  echo file changed
done
```

3. Don't forget to `chmod +x /tmp/bin/awesome-test.sh`.

If awesome was started with something like `startx` on a tty you should see the process `/tmp/bin/shell-script-with-pipe.sh` in your processes viewer with `inotifywait /tmp/awesome-test` as a child process.

**Actual result:**

After exiting awesome, when looking at `top`, a process `inotifywait /tmp/awesome-test` should still be present.

If someone is running such a script asynchronously in his awesome session, he won't actually be logged out when logging out of the tty from which awesome was spawned. His home folder won't be decrypted and the process will stay there stale..

**Expected result:**

The whole process' subprocesses are all killed.

**Comment by psychon** [comment]
How about replacing `awesome.kill(pid, awesome.unix_signal['SIGTERM'])` with `awesome.kill(-pid, awesome.unix_signal['SIGTERM'])` (note the extra minus sign). This requires commit b3311674d2073a0fdea35f033dcc06d6373d4873. If you do this, then add a comment explaing that

> killing -p means sending a signal to every process in the process group p. Awesome makes sure to spawn processes in a new session, so this works.

Alternatively, if you do not have a version of awesome with this commit, you could try `awful.spawn{"kill", tostring(-pid)}` to kill the process group.

> Expected result:
>
> The whole process' subprocesses are all killed.

Note that AFAIK Unix does not actually allow to get this expected result. On Linux it can be done with control groups, but in general a process can escape by forking twice and calling `setsid()`. The intermediate process then exits and the parent process id of the remaining process becomes pid 1.

**Comment by doronbehar** [comment]
First of all, thanks for your quick response! This should be noted in the documentation, I'm glad it is supposed to have a workaround already. I'll patch and compile awesome and use it until the next release.

Just out of curiosity, why isn't killing the process group is the default behavior?

**Comment by psychon** [comment]
> why isn't killing the process group is the default behavior?

Ask the inventors of Unix. Awesome just provides access to the `kill` library function. It does not define its semantics.

Does that mean this issue can be closed?

**Comment by doronbehar** [comment]
Cool, well I assume you guys have tested this and it works so I won't hold you with this issue open until I'll find the time to compile Awesome from source.

Thanks!

**Comment by Perlovka** [comment]
@doronbehar, try to add the following to your shell script:

```shell
trap killgroup INT

killgroup() {
    echo " Interrupted by user... quitting."
    kill 0
}
```
This will kill all subshells on interrupt signal.

P.S. It is possible to use inotify directly from awesome ;-P
I'll upload my module to github repo, when it's done.

**Comment by blueyed** [comment]
b3311674d (in master) allows for 0 already.

**Comment by doronbehar** [comment]
Gee, A native inotify lua module for awesome could be awesome :] I've encountered https://github.com/hoelzro/linotify when I have investigated this matter but I've reached the conclusion it will work synchronously so it won't fit my needs.

I've just finished compiling it with the patch and it works like a charm. Thanks a lot again!

That workaround to the shell script didn't work BTW, I've tested it on a different script which produced the same behavior..

**Comment by Perlovka** [comment]
@doronbehar this workaround trap INT signal, which is triggered with Ctrl+C in foreground running script. Try to use `trap killgroup TERM` or `trap killgroup KILL`

**Comment by psychon** [comment]
@blueyed However, "kill 0" means "kill my whole process group". If you do that in awesome, awesome would kill itself and would *not* kill the shell script. If you do that in the shell script, the shell script kills itself and its children.

**Comment by doronbehar** [comment]
This is included in v4.3 so closing.
